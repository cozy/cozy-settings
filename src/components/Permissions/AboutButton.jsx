import React from 'react'
import CircleButton from 'cozy-ui/transpiled/react/CircleButton'
import Icon from 'cozy-ui/transpiled/react/Icon'
import InfoIcon from 'cozy-ui/transpiled/react/Icons/Info'
import AppLinker, { generateWebLink } from 'cozy-ui/transpiled/react/AppLinker'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import { useClient } from 'cozy-client'

export const AboutButton = ({ matchingQueryResultData }) => {
  const { t } = useI18n()
  const client = useClient()
  const { subdomain: subDomainType } = client.getInstanceOptions()
  const appWebRef =
    matchingQueryResultData &&
    generateWebLink({
      cozyUrl: client.getStackClient().uri,
      slug: 'store',
      subDomainType,
      nativePath: `discover/${matchingQueryResultData.slug}`
    })
  const appData = matchingQueryResultData && matchingQueryResultData

  return (
    <AppLinker app={appData} href={appWebRef}>
      {({ onClick, href }) => (
        <CircleButton
          label={t('Permissions.about')}
          href={href}
          onClick={onClick}
        >
          <Icon icon={InfoIcon} />
        </CircleButton>
      )}
    </AppLinker>
  )
}
