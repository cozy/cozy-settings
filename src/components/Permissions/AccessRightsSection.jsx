import React from 'react'
import { NavigationListSection } from 'cozy-ui/transpiled/react/NavigationList'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import Typography from 'cozy-ui/transpiled/react/Typography'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import ListItemIcon, {
  smallSize,
  mediumSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import Icon from 'cozy-ui/transpiled/react/Icon'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemSecondaryAction'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import {
  displayPermissions,
  getPermissionIconName
} from './helpers/permissionsHelper'
import { routes } from 'constants/routes'
import withAllLocales from '../../lib/withAllLocales'

const AccessRightsSection = ({
  sortedPermissionsByName,
  slugName,
  t,
  isRemoteDoctypes
}) => {
  if (!sortedPermissionsByName || sortedPermissionsByName.length < 1) {
    return
  }
  return (
    <NavigationListSection>
      <ListItem>
        <Typography variant="h5">
          {!isRemoteDoctypes
            ? t('Permissions.limited_right_access')
            : t('Permissions.exit_rights')}
        </Typography>
      </ListItem>
      <Divider />
      {sortedPermissionsByName.map(({ name, title, verbs, type }, index) => {
        const iconName = getPermissionIconName(type)
        return (
          <div key={name}>
            <ListItem
              button
              component="a"
              href={`/#${routes.appList}/${slugName}/${name}`}
            >
              <ListItemIcon>
                <Icon
                  icon={
                    require(`cozy-ui/transpiled/react/Icons/${iconName}`)
                      .default
                  }
                  size={mediumSize}
                />
              </ListItemIcon>
              <ListItemText
                primary={title}
                secondary={t(displayPermissions(verbs))}
              />
              <ListItemSecondaryAction>
                <Icon
                  icon={RightIcon}
                  size={smallSize}
                  className="u-mr-1"
                  style={{ color: 'var(--secondaryTextColor)' }}
                />
              </ListItemSecondaryAction>
            </ListItem>
            {index !== sortedPermissionsByName.length - 1 && (
              <Divider variant="inset" />
            )}
          </div>
        )
      })}
    </NavigationListSection>
  )
}

export default withAllLocales(AccessRightsSection)
