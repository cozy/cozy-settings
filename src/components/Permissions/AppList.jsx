import React from 'react'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import Icon from 'cozy-ui/transpiled/react/Icon'
import Typography from 'cozy-ui/transpiled/react/Typography'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon, {
  smallSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { routes } from 'constants/routes'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemSecondaryAction'
import NavigationList, {
  NavigationListSection
} from 'cozy-ui/transpiled/react/NavigationList'
import { isQueryLoading, hasQueryBeenLoaded } from 'cozy-client'
import { buildAppsQuery, buildKonnectorsQuery } from 'lib/queries'
import { useQuery } from 'cozy-client'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

const AppList = () => {
  const { t } = useI18n()
  const appsQuery = buildAppsQuery()
  const queryResultApps = useQuery(appsQuery.definition, appsQuery.options)
  const konnectorsQuery = buildKonnectorsQuery()
  const queryResultKonnectors = useQuery(
    konnectorsQuery.definition,
    konnectorsQuery.options
  )
  const { isMobile, isTablet } = useBreakpoints()
  const toNotDisplay = ['home', 'store', 'settings']

  return (
    <Page
      className={isMobile || isTablet ? '' : 'u-maw-7'}
      withoutMarginTop={isMobile || isTablet}
    >
      <PageTitle>{t('Permissions.applications')}</PageTitle>
      {(isQueryLoading(queryResultApps) ||
        isQueryLoading(queryResultKonnectors)) &&
      (!hasQueryBeenLoaded(queryResultApps) ||
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' ||
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <NavigationList>
          <NavigationListSection>
            {queryResultApps.data
              .concat(queryResultKonnectors.data)
              .sort((a, b) => a.slug.localeCompare(b.slug))
              .filter(a => !toNotDisplay.includes(a.slug))
              .map(appOrKonnector => {
                return (
                  <div key={appOrKonnector.name}>
                    <ListItem
                      button
                      component="a"
                      href={`/#${routes.appList}/${appOrKonnector.slug}`}
                    >
                      <ListItemIcon>
                        <AppIcon app={appOrKonnector} />
                      </ListItemIcon>
                      <ListItemText
                        primary={appOrKonnector.name}
                        secondary={t('Permissions.numberOfPermissions', {
                          smart_count: Object.entries(
                            appOrKonnector.permissions
                          ).length
                        })}
                      />
                      <ListItemSecondaryAction>
                        <Icon
                          icon={RightIcon}
                          size={smallSize}
                          className="u-mr-1"
                          style={{ color: 'var(--secondaryTextColor)' }}
                        />
                      </ListItemSecondaryAction>
                    </ListItem>
                    <Divider variant="inset" />
                  </div>
                )
              })}
          </NavigationListSection>
        </NavigationList>
      )}
    </Page>
  )
}

export default AppList
