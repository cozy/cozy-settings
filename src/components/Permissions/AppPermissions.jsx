import React from 'react'

import { APPS_DOCTYPE, KONNECTORS_DOCTYPE } from 'doctypes'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Typography from 'cozy-ui/transpiled/react/Typography'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import { Link, useParams } from 'react-router-dom'
import {
  displayPermissions,
  getPermissionIconName
} from '../../containers/helpers/permissionsHelper'
import Icon from 'cozy-ui/transpiled/react/Icon'
import IconButton from 'cozy-ui/transpiled/react/IconButton'
import PreviousIcon from 'cozy-ui/transpiled/react/Icons/Previous'
import ListItemIcon, {
  smallSize,
  mediumSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import { routes } from 'constants/routes'
import { OpenappButton } from './OpenappButton'
import withAllLocales from '../../lib/withAllLocales'
import ListItemSecondaryAction from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemSecondaryAction'
import RightIcon from 'cozy-ui/transpiled/react/Icons/Right'
import NavigationList from 'cozy-ui/transpiled/react/NavigationList'
import CozyClient, {
  Q,
  useQuery,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'

export const completePermission = (
  name,
  permission,
  { description, verbs },
  type
) => {
  const completedPermission = {
    name,
    type,
    title: permission
  }
  if (description) {
    completedPermission.description = description
  }
  if (verbs) {
    completedPermission.verbs = verbs
  }
  return completedPermission
}

const PermissionsApplication = ({ t }) => {
  const { slug: slugName } = useParams()
  const THIRTY_SECONDS = 30 * 1000

  const queryResultApps = useQuery(
    Q(APPS_DOCTYPE).getById('io.cozy.apps/' + slugName),
    {
      as: 'io.cozy.apps/' + slugName,
      fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
    }
  )

  const queryResultKonnectors = useQuery(
    Q(KONNECTORS_DOCTYPE).getById('io.cozy.konnectors/' + slugName),
    {
      as: 'io.cozy.konnectors/' + slugName,
      fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
    }
  )

  let matchingQueryResult
  if (queryResultApps?.data?.length > 0) {
    matchingQueryResult = queryResultApps
  }
  if (queryResultKonnectors?.data?.length > 0) {
    matchingQueryResult = queryResultKonnectors
  }
  const sortPermissionsByName = queryResult => {
    return Object.entries(queryResult.data[0].attributes.permissions)
      .map(([name, value]) => {
        const type = value.type
        const perm = t('CozyPermissions.Permissions.' + type)
        return completePermission(name, perm, value, type)
      })
      .sort((a, b) => {
        return a.title.localeCompare(b.title)
      })
  }

  const isKonnector = type => type === 'io.cozy.konnectors'

  let sortedPermissionsByName =
    matchingQueryResult && sortPermissionsByName(matchingQueryResult)

  const isNotLastItem = name => {
    return (
      name !== sortedPermissionsByName[sortedPermissionsByName.length - 1].name
    )
  }

  return (
    <Page narrow>
      {(isQueryLoading(queryResultApps) &&
        !hasQueryBeenLoaded(queryResultApps)) ||
      (isQueryLoading(queryResultKonnectors) &&
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' &&
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <div>
          <IconButton className="u-mr-half" href="#/permissions">
            <Icon icon={PreviousIcon} size={16} />
          </IconButton>
          <NavigationList>
            <PageTitle>{slugName.toUpperCase()}</PageTitle>
            <OpenappButton
              type={
                isKonnector(matchingQueryResult.data[0].type)
                  ? 'konnector'
                  : 'app'
              }
              matchingQueryResultData={matchingQueryResult.data[0]}
            />

            {sortedPermissionsByName.map(({ name, title, verbs, type }) => {
              const iconName = getPermissionIconName(type)
              return (
                <Link
                  to={`${routes.appList}/${slugName}/${name}`}
                  key={name}
                  style={{ textDecoration: 'none', color: 'black' }}
                >
                  <ListItem button>
                    <ListItemIcon>
                      <Icon
                        icon={
                          require(`cozy-ui/transpiled/react/Icons/${iconName}`)
                            .default
                        }
                        size={mediumSize}
                      />
                    </ListItemIcon>
                    <ListItemText
                      primary={title}
                      secondary={t(displayPermissions(verbs))}
                    />
                    <ListItemSecondaryAction style={{ color: 'grey' }}>
                      <Icon
                        icon={RightIcon}
                        size={smallSize}
                        className="u-mr-1"
                      />
                    </ListItemSecondaryAction>
                  </ListItem>
                  {isNotLastItem(name) && <Divider variant="inset" />}
                </Link>
              )
            })}
          </NavigationList>
        </div>
      )}
    </Page>
  )
}

export default withAllLocales(PermissionsApplication)
