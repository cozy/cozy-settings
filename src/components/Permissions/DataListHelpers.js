export const completePermission = (apps, konnectors) => {
  const hiddenSlugs = ['home', 'store', 'settings']
  let slugs = []
  const permissionsToDisplay = {}
  if (apps?.data?.length > 0 || konnectors?.data?.length > 0) {
    slugs = [
      ...apps.data.filter(({ slug }) => !hiddenSlugs.includes(slug)),
      ...konnectors.data
    ]

    for (const { slug, permissions } of slugs) {
      Object.values(permissions).map(({ type }) => {
        if (!permissionsToDisplay[type]) {
          permissionsToDisplay[type] = []
        }
        permissionsToDisplay[type].push(slug)
      })
    }
  }
  return permissionsToDisplay
}

export const sortPermissionsByName = (permissionsToDisplay, t) => {
  return Object.entries(permissionsToDisplay).sort((a, b) => {
    return t('CozyPermissions.Permissions.' + a[0])
      .toString()
      .localeCompare(t('CozyPermissions.Permissions.' + b[0]))
  })
}
