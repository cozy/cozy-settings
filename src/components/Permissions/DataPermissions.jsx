import React from 'react'

import { APPS_DOCTYPE, KONNECTORS_DOCTYPE } from 'doctypes'
import { Link, useParams } from 'react-router-dom'
import withAllLocales from 'lib/withAllLocales'
import Typography from 'cozy-ui/transpiled/react/Typography'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import NavigationList from 'cozy-ui/transpiled/react/NavigationList'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import PreviousIcon from 'cozy-ui/transpiled/react/Icons/Previous'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Icon from 'cozy-ui/transpiled/react/Icon'
import IconButton from 'cozy-ui/transpiled/react/IconButton'
import { routes } from 'constants/routes'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import { retrieveSlugsOfPermission } from './helpers/DataPermissionsHelpers'

import CozyClient, {
  Q,
  useQuery,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'

const DataPermissions = ({ t }) => {
  const { permission: permissionType } = useParams()
  const THIRTY_SECONDS = 30 * 1000

  const queryResultApps = useQuery(Q(APPS_DOCTYPE), {
    as: APPS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })
  const queryResultKonnectors = useQuery(Q(KONNECTORS_DOCTYPE), {
    as: KONNECTORS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })

  const slugs = retrieveSlugsOfPermission(
    queryResultApps,
    queryResultKonnectors,
    permissionType
  )

  return (
    <Page narrow>
      <IconButton className="u-mr-half" href={`#${routes.dataList}`}>
        <Icon icon={PreviousIcon} size={16} />
      </IconButton>
      <PageTitle>
        {t('CozyPermissions.Permissions.' + permissionType)}
      </PageTitle>
      {(isQueryLoading(queryResultApps) &&
        !hasQueryBeenLoaded(queryResultApps)) ||
      (isQueryLoading(queryResultKonnectors) &&
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' &&
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <NavigationList>
          {slugs.map(slug => {
            return (
              <React.Fragment key={slug.slug}>
                <Link to="" style={{ textDecoration: 'none' }}>
                  <ListItem button>
                    <ListItemIcon>
                      <AppIcon app={slug} />
                    </ListItemIcon>
                    <ListItemText primary={slug.slug} />
                  </ListItem>
                </Link>
                <Divider />
              </React.Fragment>
            )
          })}
        </NavigationList>
      )}
    </Page>
  )
}

export default withAllLocales(DataPermissions)
