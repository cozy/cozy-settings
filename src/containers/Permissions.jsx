import React from 'react'
import { Link } from 'react-router-dom'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import Typography from 'cozy-ui/transpiled/react/Typography'
import { useI18n } from 'cozy-ui/transpiled/react'
import List from 'cozy-ui/transpiled/react/MuiCozyTheme/List'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Spinner from 'cozy-ui/transpiled/react/Spinner'

import { APPS_DOCTYPE, KONNECTORS_DOCTYPE } from 'doctypes'
import CozyClient, {
  Q,
  useQuery,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'

const Permissions = () => {
  const { t } = useI18n()
  const THIRTY_SECONDS = 30 * 1000
  const queryResultApps = useQuery(Q(APPS_DOCTYPE), {
    as: APPS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })
  const queryResultKonnectors = useQuery(Q(KONNECTORS_DOCTYPE), {
    as: KONNECTORS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })

  return (
    <Page narrow>
      <PageTitle>{t('Permissions.title')}</PageTitle>
      {(isQueryLoading(queryResultApps) ||
        isQueryLoading(queryResultKonnectors)) &&
      (!hasQueryBeenLoaded(queryResultApps) ||
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' ||
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <List>
          {queryResultApps.data
            .concat(queryResultKonnectors.data)
            .sort((a, b) => a.slug.localeCompare(b.slug))
            .map(appOrKonnector => {
              return (
                <div key={appOrKonnector.name}>
                  <Link to={'/permissions/' + appOrKonnector.slug}>
                    <ListItem button>
                      <ListItemIcon>
                        <AppIcon app={appOrKonnector} />
                      </ListItemIcon>
                      <ListItemText
                        primary={appOrKonnector.name}
                        secondary={t('Permissions.numberOfPermissions', {
                          smart_count: Object.entries(
                            appOrKonnector.permissions
                          ).length
                        })}
                      />
                    </ListItem>
                  </Link>
                  <Divider />
                </div>
              )
            })}
        </List>
      )}
    </Page>
  )
}

export default Permissions
