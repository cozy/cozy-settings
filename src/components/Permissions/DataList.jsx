import React from 'react'
import { Link } from 'react-router-dom'
import Page from 'components/Page'
import PageTitle from 'components/PageTitle'
import Typography from 'cozy-ui/transpiled/react/Typography'
import NavigationList, {
  NavigationListSection
} from 'cozy-ui/transpiled/react/NavigationList'
import ListItem from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItem'
import ListItemText from 'cozy-ui/transpiled/react/ListItemText'
import ListItemIcon, {
  mediumSize
} from 'cozy-ui/transpiled/react/MuiCozyTheme/ListItemIcon'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import Icon from 'cozy-ui/transpiled/react/Icon'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { APPS_DOCTYPE, KONNECTORS_DOCTYPE } from 'doctypes'
import withAllLocales from 'lib/withAllLocales'
import { getPermissionIconName } from 'containers/helpers/permissionsHelper'
import { completePermission, sortPermissionsByName } from './DataListHelpers'
import { routes } from 'constants/routes'
import CozyClient, {
  Q,
  useQuery,
  isQueryLoading,
  hasQueryBeenLoaded
} from 'cozy-client'

const DataList = ({ t }) => {
  const THIRTY_SECONDS = 30 * 1000
  const queryResultApps = useQuery(Q(APPS_DOCTYPE), {
    as: APPS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })
  const queryResultKonnectors = useQuery(Q(KONNECTORS_DOCTYPE), {
    as: KONNECTORS_DOCTYPE,
    fetchPolicy: CozyClient.fetchPolicies.olderThan(THIRTY_SECONDS)
  })

  const permissionsToDisplay = completePermission(
    queryResultApps,
    queryResultKonnectors
  )

  return (
    <Page narrow>
      <PageTitle>{t('Permissions.data')}</PageTitle>
      {(isQueryLoading(queryResultApps) ||
        isQueryLoading(queryResultKonnectors)) &&
      (!hasQueryBeenLoaded(queryResultApps) ||
        !hasQueryBeenLoaded(queryResultKonnectors)) ? (
        <Spinner size="large" className="u-flex u-flex-justify-center u-mt-1" />
      ) : queryResultApps.fetchStatus === 'failed' ||
        queryResultKonnectors.fetchStatus === 'failed' ? (
        <Typography variant="body1" className="u-mb-1-half">
          {t('Permissions.failedRequest')}
        </Typography>
      ) : (
        <NavigationList>
          <NavigationListSection>
            {sortPermissionsByName(permissionsToDisplay, t).map(
              ([key, slugs]) => {
                const iconName = getPermissionIconName(key)
                return (
                  <React.Fragment key={key}>
                    <Link to={`${routes.dataList}/${key}`}>
                      <ListItem button>
                        <ListItemIcon>
                          <Icon
                            icon={
                              require(`cozy-ui/transpiled/react/Icons/${iconName}`)
                                .default
                            }
                            size={mediumSize}
                          />
                        </ListItemIcon>
                        <ListItemText
                          primary={t('CozyPermissions.Permissions.' + key)}
                          secondary={t('Permissions.numberOfApplications', {
                            smart_count: slugs.length
                          })}
                        />
                      </ListItem>
                    </Link>
                    <Divider />
                  </React.Fragment>
                )
              }
            )}
          </NavigationListSection>
        </NavigationList>
      )}
    </Page>
  )
}

export default withAllLocales(DataList)
