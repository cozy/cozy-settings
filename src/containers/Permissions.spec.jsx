import { render } from '@testing-library/react'
import React from 'react'
import Permissions from './Permissions'
import { Q, useQuery, isQueryLoading, hasQueryBeenLoaded } from 'cozy-client'

jest.mock('cozy-ui/transpiled/react', () => {
  return { useI18n: () => ({ t: x => x }) }
})

jest.mock('react-router-dom', () => {
  return {
    Link: ({ narrow, children }) => (
      <div data-testid="page" data-narrow={narrow}>
        {children}
      </div>
    )
  }
})

jest.mock('cozy-client')

jest.mock('components/Page', () => {
  // eslint-disable-next-line react/display-name
  return ({ narrow, children }) => (
    <div data-testid="page" data-narrow={narrow}>
      {children}
    </div>
  )
})

jest.mock('components/PageTitle', () => {
  // eslint-disable-next-line react/display-name
  return ({ children }) => <div data-testid="page-title">{children}</div>
})

jest.mock('cozy-ui/transpiled/react/Spinner', () => {
  // eslint-disable-next-line react/display-name
  return ({ size }) => <div data-testid="Spinner" data-size={size}></div>
})

jest.mock('cozy-ui/transpiled/react/AppIcon', () => {
  // eslint-disable-next-line react/display-name
  return () => <div data-testid="AppIcon"></div>
})

describe('Permissions', () => {
  beforeEach(() => {
    const queryResultApps = {
      fetchStatus: 'loaded',
      data: [
        {
          attributes: {
            permissions: {
              files: {
                type: 'io.cozy.files',
                description: 'Required to access the files'
              },
              allFiles: {
                type: 'io.cozy.files.*',
                description: 'Required to access the files'
              },
              apps: {
                type: 'io.cozy.apps',
                description:
                  'Required by the cozy-bar to display the icons of the apps',
                verbs: ['GET']
              }
            }
          },
          slug: 'contacts',
          name: 'Contacts'
        }
      ]
    }
    const queryResultKonnectors = {
      fetchStatus: 'loaded',
      data: [
        {
          attributes: {
            permissions: {
              files: {
                type: 'io.cozy.files',
                description: 'Required to access the files'
              },
              allFiles: {
                type: 'io.cozy.files.*',
                description: 'Required to access the files'
              },
              konnectors: {
                description: 'Required to get the list of konnectors',
                type: 'io.cozy.konnectors',
                verbs: ['GET']
              }
            }
          },
          slug: 'alan',
          name: 'Alan'
        }
      ]
    }
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(true)
    Q.mockReturnValue({ getById: () => 'kfrf' })
    useQuery.mockReturnValueOnce(queryResultApps)
    useQuery.mockReturnValueOnce(queryResultKonnectors)
  })
  it('should display appName when query has been loaded', () => {
    hasQueryBeenLoaded.mockReturnValue(true)
    const { container } = render(<Permissions />)
    expect(container).toMatchSnapshot()
  })

  it('should render a spinner when query is loading and has not been loaded', () => {
    isQueryLoading.mockReturnValue(true)
    hasQueryBeenLoaded.mockReturnValue(false)
    const { queryByTestId } = render(<Permissions />)
    expect(queryByTestId('Spinner')).toBeTruthy()
  })

  it('should display appName and konnectorName when query is not loading', () => {
    isQueryLoading.mockReturnValue(false)
    const { queryByText } = render(<Permissions />)
    expect(queryByText('Contacts')).toBeTruthy()
    expect(queryByText('Alan')).toBeTruthy()
  })

  it('should render Permissions.failedRequest when query status is failed', () => {
    useQuery.mockReset()
    useQuery.mockReturnValue({ fetchStatus: 'failed' })
    const { queryByText } = render(<Permissions />)
    expect(queryByText('Permissions.failedRequest')).toBeTruthy()
  })
})
