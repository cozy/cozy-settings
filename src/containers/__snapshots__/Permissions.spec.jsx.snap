// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permissions should display appName when query has been loaded 1`] = `
<div>
  <div
    data-narrow="true"
    data-testid="page"
  >
    <div
      data-testid="page-title"
    >
      Permissions.title
    </div>
    <ul
      class="MuiList-root MuiList-padding"
    >
      <div>
        <div
          data-testid="page"
        >
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <div
                data-testid="AppIcon"
              />
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root u-db u-ellipsis MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Alan
              </span>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </div>
        <hr
          class="MuiDivider-root"
        />
      </div>
      <div>
        <div
          data-testid="page"
        >
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <div
                data-testid="AppIcon"
              />
            </div>
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root u-db u-ellipsis MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Contacts
              </span>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </div>
        <hr
          class="MuiDivider-root"
        />
      </div>
    </ul>
  </div>
</div>
`;
